tools/python/boutiques/localExec.py

```diff
@@ -1155,6 +1177,112 @@ def check(keyname, isGood, msg, value):  # Checks input values
                 message += ("\t" + err + "\n")
             raise_error(ExecutorError, message)
 
+    # Private method to generate summary object of data
+    # collection file containing the tool name and descriptor DOI
+    def _generateSummary(self, desc):
+        summary = {}
+        summary['name'] = self.desc_dict['name']
+        summary['descriptor-doi'] = self._findDOI(desc)
+        return summary
+
+    # Private method to attempt to find descriptor DOI
+    # through various cases
+    # userInput may be json string, filename or zenodo id
+    def _findDOI(self, userInput):
+        # DOI in Zenodo reference case
+        if userInput.split(".")[0].lower() == "zenodo":
+            return userInput
+        # File cases
+        if os.path.isfile(userInput):
+            # Most recent DOI in file if user is publisher
+            # Include check to ensure descriptor is unmodified
+            if self.desc_dict.get('doi') is not None:
+                doi = self.desc_dict.get('doi')
+                if loadJson(doi) == self.desc_dict:
+                    return doi
+            # DOI in filename if descriptor pulled from Zenodo
+            # Include check to ensure descriptor is as published
+            elif os.path.basename(userInput).split("-")[0].lower() == "zenodo":
+                doi = os.path.basename(userInput).split(".")[0].replace("-",".")
+                if loadJson(doi) == self.desc_dict:
+                    return doi
+        # TODO: No DOI found, Handle descriptor must be published case
+        return None
+
+    # Private method to generate public invocation object for data collection file
+    # absolute paths are stripped to filenames and hashes are generated for
+    # each input of type File, all other inputs are recorded as submitted
+    def _generatePublicInvocation(self):
+        public_in_dict = self.in_dict.copy()
+        # Replace file type inputs with object containing input file hash and filename
+        for x in self.inputs:
```

Comment:
you may want to turn this for loop into dictionary comprehension to make it more efficient.

Reply:
I attempted this but found there were too many pieces in play here and feel that the readability of the code is much improved in the loop form.